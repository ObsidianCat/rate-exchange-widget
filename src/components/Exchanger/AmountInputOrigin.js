import { InputNumber } from 'antd';
import React from "react";
import { connect } from "react-redux";
import {updateOriginAmount} from "../../actions";

class AmountInputOrigin extends React.Component{

  render(){
    return (
      <InputNumber min={0} value={this.props.exchange.amount} onChange={this.props.updateOriginAmount} />
    )
  }
}

const mapStateToProps = ({exchange }) => ({
  exchange
});

export default connect(
  mapStateToProps,
  { updateOriginAmount }
)(AmountInputOrigin);