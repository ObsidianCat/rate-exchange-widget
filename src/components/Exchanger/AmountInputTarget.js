import React from "react";
import { InputNumber } from "antd";
import { connect } from "react-redux";
import { updateOriginAmount } from "../../actions";
import {
  calculateTargetFromOrigin,
  calculateOriginFromTarget
} from "../../utils";

class AmountInputTarget extends React.Component {
  calculateOrigin = target => {
    const origin = calculateOriginFromTarget(
      target,
      this.props.exchangeRates,
      this.props.exchange.targetCurrency
    );
    this.props.updateOriginAmount(origin);
  };

  render() {
    const { exchange, exchangeRates } = this.props;
    const result = calculateTargetFromOrigin(
      exchange.amount,
      exchangeRates,
      exchange.targetCurrency
    );
    return (
      <InputNumber
        min={0}
        value={isNaN(result) ? 0 : result}
        onChange={this.calculateOrigin}
      />
    );
  }
}

const mapStateToProps = ({ exchange, exchangeRates }) => ({
  exchange,
  exchangeRates
});

export default connect(
  mapStateToProps,
  { updateOriginAmount }
)(AmountInputTarget);
