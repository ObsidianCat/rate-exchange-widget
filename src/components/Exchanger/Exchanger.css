.exchanger{
  width: 400px;
  margin: 10px auto 10px auto;
}

.exchanger .ant-card-body{
  margin: 0;
  padding: 0;
}

.origin, .target{
  display: flex;
  justify-content: center;
  align-items: baseline;
  padding: 20px 5px;
}
.target {
  background-color: #0000000f;
}

.target input:hover{
  border: 1px solid #d9d9d9;
}

.origin > *, .target > * {
  margin: 10px;
}

.swap-btn{
  position: absolute;
  margin-top: -10px;
  left: 10px;
}