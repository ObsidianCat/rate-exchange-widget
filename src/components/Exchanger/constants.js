export const CURRENCY_ORIGIN = "CURRENCY_ORIGIN";
export const CURRENCY_TARGET = "CURRENCY_TARGET";
