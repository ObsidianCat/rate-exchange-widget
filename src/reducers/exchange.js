import {
  SET_ORIGIN_CURRENCY_TYPE,
  SET_TARGET_CURRENCY_TYPE,
  SET_AMOUNT,
  SWAP_CURRENCY_TYPE
} from "../actions/actionTypes";

export const defaultState = {
  originCurrency: "EUR",
  targetCurrency: "USD",
  amount: 50
};

export default function exchange(state = defaultState, action) {
  switch (action.type) {
    case SWAP_CURRENCY_TYPE:
      return Object.assign({}, state, {
        originCurrency: state.targetCurrency,
        targetCurrency: state.originCurrency
      });
    case SET_ORIGIN_CURRENCY_TYPE:
      return Object.assign({}, state, { originCurrency: action.payload });
    case SET_TARGET_CURRENCY_TYPE:
      return Object.assign({}, state, { targetCurrency: action.payload });
    case SET_AMOUNT:
      return Object.assign({}, state, { amount: action.payload });
    default:
      return state;
  }
}
