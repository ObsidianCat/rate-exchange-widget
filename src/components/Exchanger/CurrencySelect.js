import React from "react";
import { Select } from "antd";
import { connect } from "react-redux";
import { setCurrencyTypeTarget, setCurrencyTypeOrigin } from "../../actions";
import { CURRENCY_ORIGIN } from "./constants";

const Option = Select.Option;

class CurrencySelect extends React.Component {
  handleCurrencyChange = value => {
    if (this.props.curType === CURRENCY_ORIGIN) {
      this.props.setCurrencyTypeOrigin(value, this.props.curType);
    } else {
      this.props.setCurrencyTypeTarget(value, this.props.curType);
    }
  };

  render() {
    const options = [];
    for (let prop in this.props.exchangeRates) {
      options.push(
        <Option value={prop} key={prop}>
          {prop}
        </Option>
      );
    }

    const selected =
      this.props.curType === CURRENCY_ORIGIN
        ? this.props.exchange.originCurrency
        : this.props.exchange.targetCurrency;

    return (
      <Select
        value={selected}
        style={{ width: 120 }}
        onChange={this.handleCurrencyChange}
      >
        {options}
      </Select>
    );
  }
}

const mapStateToProps = ({ exchange }) => ({
  exchange
});

export default connect(
  mapStateToProps,
  { setCurrencyTypeOrigin, setCurrencyTypeTarget }
)(CurrencySelect);
