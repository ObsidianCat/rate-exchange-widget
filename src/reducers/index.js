import { combineReducers } from "redux";
import exchangeRates from "./exchangeRates";
import exchange from "./exchange";

export default combineReducers({
  exchangeRates,
  exchange
});
