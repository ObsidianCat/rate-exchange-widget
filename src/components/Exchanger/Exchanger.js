import React from "react";
import { connect } from "react-redux";
import { fetchExchangeRates, swapCurrencyType, doSwap } from "../../actions";
import CurrencySelect from "./CurrencySelect";
import AmountInputOrigin from "./AmountInputOrigin";
import AmountInputTarget from "./AmountInputTarget";
import { CURRENCY_ORIGIN, CURRENCY_TARGET } from "./constants";
import { Button } from "antd";
import "./Exchanger.css";
import { Card } from "antd";

export class ExchangerComponent extends React.Component {
  componentDidMount() {
    this.props.fetchExchangeRates();
    this.fetcherRef = setInterval(this.props.fetchExchangeRates, 100000);
  }

  componentWillUnmount() {
    clearInterval(this.fetcherRef);
  }

  render() {
    return (
      <Card className="exchanger">
        <div className="origin">
          <CurrencySelect
            exchangeRates={this.props.exchangeRates}
            selected={this.props.exchange.originCurrency}
            curType={CURRENCY_ORIGIN}
          />
          <AmountInputOrigin />
        </div>

        <div>
          <Button
            type="primary"
            icon="swap"
            size={"small"}
            onClick={this.props.doSwap}
            className="swap-btn"
          >
            Swap
          </Button>
        </div>
        <div className="target">
          <CurrencySelect
            exchangeRates={this.props.exchangeRates}
            selected={this.props.exchange.targetCurrency}
            curType={CURRENCY_TARGET}
          />
          <AmountInputTarget
            originAmount={this.props.exchange.amount}
            exchangeRate={
              this.props.exchangeRates[this.props.exchange.targetCurrency]
            }
          />
        </div>
      </Card>
    );
  }
}

const mapStateToProps = ({ exchangeRates, exchange }) => ({
  exchangeRates,
  exchange
});

export default connect(
  mapStateToProps,
  { fetchExchangeRates, swapCurrencyType, doSwap }
)(ExchangerComponent);
