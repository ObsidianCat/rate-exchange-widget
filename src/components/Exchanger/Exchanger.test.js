import React from 'react';
import { shallow } from 'enzyme';
import { ExchangerComponent } from './Exchanger';
import {defaultState as exchangeState} from "../../reducers/exchange";

let wrapper;

const mockProps = {
  exchange: exchangeState,
  exchangeRates: {},
  fetchExchangeRates:jest.fn(),
  swapCurrencyType: jest.fn(),
  doSwap: jest.fn()
}

beforeAll(()=>{
  wrapper =  shallow(<ExchangerComponent {...mockProps}/>);
})
afterAll(()=>{
  wrapper.unmount()
})

it('should render as snapshot', () => {
  console.log(wrapper.props)
  expect(wrapper).toMatchSnapshot();
});

it('should request exchange rates after mount', () => {
  wrapper.instance().componentDidMount()
  expect(mockProps.fetchExchangeRates).toBeCalled()
});

it('should do swap', () => {
  wrapper.find('.swap-btn').simulate('click')
  expect(mockProps.doSwap).toBeCalled()
});

