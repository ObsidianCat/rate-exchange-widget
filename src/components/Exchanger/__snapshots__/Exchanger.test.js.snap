// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as snapshot 1`] = `
<Card
  className="exchanger"
>
  <div
    className="origin"
  >
    <Connect(CurrencySelect)
      curType="CURRENCY_ORIGIN"
      exchangeRates={Object {}}
      selected="EUR"
    />
    <Connect(AmountInputOrigin) />
  </div>
  <div>
    <Button
      block={false}
      className="swap-btn"
      ghost={false}
      icon="swap"
      loading={false}
      onClick={[MockFunction]}
      prefixCls="ant-btn"
      size="small"
      type="primary"
    >
      Swap
    </Button>
  </div>
  <div
    className="target"
  >
    <Connect(CurrencySelect)
      curType="CURRENCY_TARGET"
      exchangeRates={Object {}}
      selected="USD"
    />
    <Connect(AmountInputTarget)
      originAmount={50}
    />
  </div>
</Card>
`;
